macro-fullcalendar-api/src/main/java/org/xwiki/fullcalendar/internal/DefaultFullCalendarManager.java

```diff
@@ -88,143 +79,222 @@ public String iCalToJSON(String iCalStringURL) throws Exception
-                .getDates(event.getStartDate().getDate(), icalIntervalStart, icalIntervalEnd, Value.DATE_TIME);
-            String groupId = String.format("%s_group", jsonMap.get(JSON_KEY_ID));
-            for (int i = 0; i < recurringEventStartDates.size(); i++) {
-                if (recurringEventStartDates.get(i).equals(event.getStartDate().getDate())) {
-                    continue;
-                }
-                Map<String, Object> recurringEvent = new HashMap<>(jsonMap);
-
-                recurringEvent.put(JSON_KEY_START_DATE,
-                    jsonDateFormat.format(recurringEventStartDates.get(i)));
-                recurringEvent.put(JSON_KEY_END_DATE, jsonDateFormat.format(
-                    new DateTime(recurringEventStartDates.get(i).getTime() + differenceInMillis)));
-                recurringEvent.put(JSON_KEY_ID, String.format("%s_%d", jsonMap.get(JSON_KEY_ID), i));
-                recurringEvent.put("groupId", groupId);
-                jsonArrayList.add(recurringEvent);
+        String groupId = String.format(GROUP_ID_FORMAT, jsonMap.getId());
+        for (int i = 0; i < recurringEventStartDates.size(); i++) {
+            if (recurringEventStartDates.get(i).equals(event.getStartDate().getDate())) {
+                continue;
             }
+            MoccaCalendarEvent recurringEvent = new MoccaCalendarEvent(jsonMap);
+
+            recurringEvent.setStart(recurringEventStartDates.get(i));
+            recurringEvent.setEnd(new DateTime(recurringEventStartDates.get(i).getTime() + differenceInMillis));
+            recurringEvent.setId(String.format("%s_%d", jsonMap.getId(), i));
+            recurringEvent.setGroupId(groupId);
+            jsonArrayList.add(recurringEvent);
         }
     }
 
-    private boolean areIntervalsIntersected(Date intervalStart1, Date intervalEnd1, Date intervalStart2,
-        Date intervalEnd2)
+    private void addRecurringEventsCollapsed(MoccaCalendarEvent jsonMap, TimeZone timeZone,
+        ArrayList<MoccaCalendarEvent> jsonArrayList, Recur recur)
     {
-        return (intervalEnd1.after(intervalStart2) && intervalEnd1.before(intervalEnd2))
-            || (intervalStart1.after(intervalStart2) && intervalStart1.before(intervalEnd2));
+        String groupId = String.format(GROUP_ID_FORMAT, jsonMap.getId());
+
+        if (recur.getUntil() != null) {
+            jsonMap.setRecEndDate(recur.getUntil());
+        } else {
+            LocalDate localDateStartDate = jsonMap.getStart().toInstant().atZone(timeZone.toZoneId()).toLocalDate();
+            LocalDate futureLocalDate = localDateStartDate.plusYears(5);
+            jsonMap.setRecEndDate(java.util.Date.from(futureLocalDate.atStartOfDay(timeZone.toZoneId()).toInstant()));
+        }
+        jsonMap.setRecurrent(1);
+        String frequency = recur.getFrequency().name();
+        String eventFrequency = getEventFrequency(recur, frequency);
+        jsonMap.setRecurrenceFreq(eventFrequency);
```

Comment:
The reccurence rule of the ical standard is can be quite complex https://icalendar.org/iCalendar-RFC-5545/3-8-5-3-recurrence-rule.html . Saving only the frequency leaves out much information. We should consider improving our application in this regard in the future.

Reply:
I agree that some improvements in this regard can be done in the future, as currently we only take into consideration the frequency.